File: Kafka/Lab3/producer.py
import csv
import json
import sys
from confluent_kafka import SerializingProducer
from confluent_kafka.serialization import StringSerializer

def delivery_report(err, msg):
    """ Called once for each message produced to indicate delivery result. """
    if err is not None:
        print(f"Message delivery failed: {err}")
    else:
        # For SerializingProducer, msg.key() and msg.value() return the serialized bytes
        print(f"Delivered to {msg.topic()} [{msg.partition()}] at offset {msg.offset()} (Key: {msg.key().decode('utf-8')})")

def main():
    # Configure the Serializing Producer
    # We define BOTH key and value to use the standard StringSerializer
    string_serializer = StringSerializer('utf_8')

    conf = {
        'bootstrap.servers': 'localhost:9092',
        'key.serializer': string_serializer,
        'value.serializer': string_serializer,
        'enable.idempotence': True
    }
    
    try:
        producer = SerializingProducer(conf)
    except Exception as e:
        print(f"Failed to create producer: {e}")
        sys.exit(1)

    topic = 'orders'
    csv_file_path = 'orders.csv'

    print(f"Starting to produce messages to topic: {topic}")

    # Read orders from CSV
    try:
        with open(csv_file_path, mode='r', encoding='utf-8') as file:
            reader = csv.DictReader(file)
            
            for row in reader:
                # Key is the order_id (string)
                key = row['order_id']
                
                # Value is the row converted into a JSON string BEFORE sending
                value = json.dumps(row)
                
                # The SerializingProducer automatically applies the StringSerializer to both!
                producer.produce(topic, key=key, value=value, on_delivery=delivery_report)
                
                # Trigger any available delivery report callbacks from previous produce() calls
                producer.poll(0)
                
    except FileNotFoundError:
        print(f"Error: {csv_file_path} not found.")
        sys.exit(1)

    # Wait for any outstanding messages to be delivered
    print("Waiting for all messages to be delivered...")
    producer.flush()
    print("Producer finished.")

if __name__ == '__main__':
    main()

File: Kafka/Lab3/Producer.java
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.clients.producer.Callback;
import org.apache.kafka.clients.producer.RecordMetadata;
import org.apache.kafka.common.serialization.StringSerializer;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Properties;

public class Producer {
    public static void main(String[] args) {
        // 1. Configure the Producer properties
        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
        // Equivalent to StringSerializer('utf_8') in Python
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());

        // 2. Create the Kafka Producer
        KafkaProducer<String, String> producer = new KafkaProducer<>(props);
        
        String topic = "orders";
        String csvFilePath = "orders.csv";

        System.out.println("Starting to produce messages to topic: " + topic);

        // 3. Read orders from the CSV file
        try (BufferedReader br = new BufferedReader(new FileReader(csvFilePath))) {
            String line;
            boolean isFirstLine = true;
            String[] headers = null;

            while ((line = br.readLine()) != null) {
                // Skip the header row but save it to build our JSON
                if (isFirstLine) {
                    headers = line.split(",");
                    isFirstLine = false;
                    continue;
                }

                String[] values = line.split(",");
                if (values.length != headers.length) continue;

                // Build a simple JSON string manually from the CSV row
                StringBuilder jsonBuilder = new StringBuilder();
                jsonBuilder.append("{");
                for (int i = 0; i < headers.length; i++) {
                    jsonBuilder.append("\"").append(headers[i]).append("\":\"").append(values[i]).append("\"");
                    if (i < headers.length - 1) {
                        jsonBuilder.append(",");
                    }
                }
                jsonBuilder.append("}");

                // Key is the order_id (first column)
                String key = values[0]; 
                String value = jsonBuilder.toString();

                // 4. Create a ProducerRecord (equivalent to producer.produce arguments)
                ProducerRecord<String, String> record = new ProducerRecord<>(topic, key, value);

                // 5. Asynchronously send the message with a delivery callback
                producer.send(record, new Callback() {
                    @Override
                    public void onCompletion(RecordMetadata metadata, Exception exception) {
                        if (exception != null) {
                            System.err.println("Message delivery failed: " + exception.getMessage());
                        } else {
                            System.out.println("Delivered to " + metadata.topic() + " [" + metadata.partition() + "] at offset " + metadata.offset() + " (Key: " + key + ")");
                        }
                    }
                });
            }
        } catch (IOException e) {
            System.err.println("Error reading " + csvFilePath + ": " + e.getMessage());
            System.exit(1);
        }

        // 6. Wait for all messages in the buffer to be delivered
        System.out.println("Waiting for all messages to be delivered...");
        producer.flush();
        producer.close();
        
        System.out.println("Producer finished.");
    }
}
